Incompatibility between DragonflyDB and redis/rueidis client

**RESP3 support missing for rueidis DragonflyDB connection?**
# Error when connecting to DragonflyDB with rueidis client

When trying to connect to a DragonflyDB instance using the rueidis Redis client library in Go, I get the following panic:

`panic: Unknown subcommand or wrong number of arguments for 'TRACKING'. Try CLIENT HELP.: ClientOption.DisableCache must be true for redis not supporting client-side caching or not supporting RESP3 [recovered] panic: Unknown subcommand or wrong number of arguments for 'TRACKING'. Try CLIENT HELP.: ClientOption.DisableCache must be true for redis not supporting client-side caching or not supporting RESP3`


This happens with the following code:

```go
package main

import (
	"context"
	"fmt"

	"github.com/redis/rueidis"
)

func main() {
	client, err := rueidis.NewClient(rueidis.ClientOption{InitAddress: []string{"127.0.0.1:6379"}})
	if err != nil {
		panic(err)
	}
	defer client.Close()

	ctx := context.Background()
	// SET key val NX
	err = client.Do(ctx, client.B().Set().Key("key").Value("val").Nx().Build()).Error()
	if err != nil {
		panic(err)
	}
	res, err := client.Do(ctx, client.B().Get().Key("key").Build()).ToString()
	if err != nil {
		panic(err)
	}
	fmt.Println(res)
}

```
It seems there may be an incompatibility between DragonflyDB and the rueidis client assumptions. Specifically the error suggests rueidis expects disabled client caching or RESP3 support.

I checked the DragonflyDB docs but didn't see mention of RESP3 support or client caching handling.

Could this be added to enable better compatibility with Redis clients like rueidis? Or is there another recommended approach for connecting to DragonflyDB from Go code?

Let me know if any other details would be helpful in resolving this. Thanks!

**Feel free to modify or improve this as needed. Hopefully it clearly explains the issue you are seeing. Let me know if you have any other questions!**

Yes, it seems that you are correct; the error appears to be different between Linux and Windows. I executed this on my MacBook and received the same error as you. let me the opportunity to double-check.

However, it seems that by adding the following parameter, it appears that the issue has been resolved:
```plaintext
DisableCache: true
```

``` GO
client, err := rueidis.NewClient(
	rueidis.ClientOption{
	InitAddress: []string{"127.0.0.1:6379"},
	DisableCache: true,
})
```
But when we set a key twice, we receive the following error:
```plaintext
panic: redis nil message
goroutine 1 [running]:
main.main()
        /Users/mozaffari/Documents/tst/dragonfly/main.go:22 +0x590
```
It seems like we might not have overwriting capabilities.

> Hi @thisismz , I run your example with go `1.21.6`. First, I am not quite sure how you are getting this output:
> 
> > 
> 
> panic: Unknown subcommand or wrong number of arguments for 'TRACKING'. Try CLIENT HELP.: ClientOption.DisableCache must be true for redis not supporting client-side caching or not supporting RESP3 [recovered] panic: Unknown subcommand or wrong number of arguments for 'TRACKING'. Try CLIENT HELP.: ClientOption.DisableCache must be true for redis not supporting client-side caching or not supporting RESP3
> 
> When I run it, I just get: `panic: syntax error`.
> 
> How I reproduced: `go mod init test && go get github.com/redis/rueidis && go run .` . The good thing is at least I managed to reproduce and figure out what the issue is. We recently added support for the `CLIENT TRACKING` subcommand in #2139 However, it appears that we only added partial support. The only tthing we accept is `CLIENT TRACKING ON/OFF`. However, `rueidis` sends `CLIENT TRACKING ON OPTIN` and dragonfly fails to parse the `OPTIN` arguments and returns with a `syntax error`.
> 
> Therefore to resolve this we need to parse the `OPTIN` argument + test with `ruedis`.
> 
> @romange what's the priority of this? Also, I noticed that we need to update the docs https://www.dragonflydb.io/docs/command-reference/compatibility to include `CLIENT TRACKING` I am mentioning it here so we don't forget

--- a/src/server/main_service.cc
+++ b/src/server/main_service.cc
@@ -1187,12 +1187,12 @@ void Service::DispatchCommand(CmdArgList args, facade::ConnectionContext* cntx) 
 
   // if this is a read command, and client tracking has enabled,
   // start tracking all the updates to the keys in this read command
-  if ((cid->opt_mask() & CO::READONLY) && dfly_cntx->conn()->IsTrackingOn()) {
+  if ((cid->opt_mask() & CO::READONLY) && dfly_cntx->conn()->IsTrackingOn() &&
+      cid->IsTransactional()) {
     auto cb = [&](Transaction* t, EngineShard* shard) {
       auto keys = t->GetShardArgs(shard->shard_id());
       return OpTrackKeys(t->GetOpArgs(shard), dfly_cntx, keys);
     };
-
     dfly_cntx->transaction->Refurbish();
     dfly_cntx->transaction->ScheduleSingleHopT(cb);
   }

--- a/src/server/server_family_test.cc
+++ b/src/server/server_family_test.cc
@@ -310,4 +310,10 @@ TEST_F(ServerFamilyTest, ClientTrackingSelectDB) {
   EXPECT_EQ(GetInvalidationMessage("IO0", 0).key, "C");
 }
 
+TEST_F(ServerFamilyTest, ClientTrackingNonTransactionalBug) {
+  Run({"HELLO", "3"});
+  Run({"CLIENT", "TRACKING", "ON"});
+
+  Run({"CLUSTER", "SLOTS"});
+}
 }  // namespace dfly

